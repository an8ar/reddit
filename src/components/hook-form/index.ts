export * from './form-provider';
export * from './rhf-text-field';
export * from './rhf-textarea';
export * from './rhf-checkbox';
