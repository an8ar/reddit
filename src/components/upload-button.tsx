import { Icon } from '@iconify/react';
import { Input } from './ui/input';
import { Label } from './ui/label';

interface Props {
  handleFileChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const UploadButton = ({ handleFileChange }: Props) => {
  return (
    <div className="absolute top-2 left-4 z-10 opacity-0 group-hover:opacity-100 transition-opacity duration-300">
      <Label
        htmlFor="inputPhoto"
        className="inset-0  px-2 py-3 rounded-full hover:cursor-pointer flex items-center gap-2 bg-stone-400 text-white"
      >
        <Icon icon="carbon:image" />
        <span>Add</span>
      </Label>
      <Input
        type="file"
        id="inputPhoto"
        accept="image/*"
        onChange={handleFileChange}
        multiple
        hidden
        className="file-selector-button-none hidden"
      />
    </div>
  );
};
