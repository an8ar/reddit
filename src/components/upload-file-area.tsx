import { Icon } from '@iconify/react/dist/iconify.js';
import { Label } from './ui/label';
import { Input } from './ui/input';
import { useTranslations } from 'next-intl';

interface Props {
  handleFileChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const FileUploadArea = ({ handleFileChange }: Props) => {
  const t = useTranslations('PostForm');

  return (
    <div className="relative w-full aspect-video rounded-2xl border-2 border-dotted border-slate-600  hover:cursor-pointer h-40 dot">
      <Label
        htmlFor="inputImage"
        className="absolute inset-0 flex gap-4 items-center justify-center"
      >
        <Icon icon="ph:plus-thin" />
        <span className="text-lg">{t('photo')}</span>
      </Label>
      <Input
        type="file"
        id="inputImage"
        accept="image/*"
        onChange={handleFileChange}
        multiple
        hidden
        className="file-selector-button-none hidden"
      />
    </div>
  );
};
