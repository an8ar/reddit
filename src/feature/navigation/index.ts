export * from './components/header';
export * from './components/sidebar';
