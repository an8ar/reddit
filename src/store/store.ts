import { combineReducers, configureStore } from "@reduxjs/toolkit";
const reducers = {};

const combinedReducers = combineReducers(reducers);
export const makeStore = () => {
  return configureStore({
    reducer: combinedReducers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {},
      }),
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];
