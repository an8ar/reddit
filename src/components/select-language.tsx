import { useLocale, useTranslations } from 'next-intl';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import * as React from 'react';

import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectLabel,
  SelectTrigger,
  SelectValue,
} from '~/components/ui/select';

export function SelectLanguage() {
  const router = useRouter();

  const t = useTranslations('LocaleSwitcher');

  const searchParams = useSearchParams().toString();

  const locale = useLocale();

  const handleLanguageChange = (value: string) => {
    router.replace(`/${value}?${searchParams}`);
  };

  return (
    <Select defaultValue={locale} onValueChange={handleLanguageChange}>
      <SelectTrigger className="w-[180px]">
        <SelectValue defaultValue={locale} />
      </SelectTrigger>
      <SelectContent>
        <SelectGroup>
          <SelectLabel>{t('title')}</SelectLabel>
          <SelectItem value="en"> {t('locale', { locale: 'en' })}</SelectItem>
          <SelectItem value="ru">{t('locale', { locale: 'ru' })}</SelectItem>
        </SelectGroup>
      </SelectContent>
    </Select>
  );
}
