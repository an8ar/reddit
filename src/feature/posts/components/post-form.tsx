"use client";

import React, { useState } from "react";
import {
  Carousel,
  CarouselContent,
  CarouselItem,
  CarouselNext,
  CarouselPrevious,
} from "~/components/ui/carousel";
import {
  FormProvider,
  RHFTextArea,
  RHFTextField,
} from "~/components/hook-form";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { FormType, Post } from "../types";
import * as Yup from "yup";
import { Button } from "~/components/ui/button";
import { useAppDispatch } from "~/store/hooks";
import { addPost } from "../posts-slice";
import Image from "next/image";
import { UploadButton } from "~/components/upload-button";
import { FileUploadArea } from "~/components/upload-file-area";
import { PostPhotoCarousel } from "./post-photo-carousel";

type FormValuesProps = Omit<Post, "id" | "createdAt">;

const CreatePostSchema = Yup.object().shape({
  title: Yup.string().required("Please fill out this field."),
  imageUrls: Yup.array().of(Yup.string().required()).optional(),
  text: Yup.string().optional(),
});

interface Props {
  closeModal?: () => void;
  type: FormType;
}

export function PostForm({ closeModal, type }: Props) {
  const [photos, setPhotos] = useState<string[]>([]);
  const dispatch = useAppDispatch();
  const defaultValues = {
    title: "",
    imageUrls: [],
  };

  const methods = useForm<FormValuesProps>({
    resolver: yupResolver(CreatePostSchema),
    defaultValues,
  });
  const { handleSubmit, setValue, getValues } = methods;

  const handleFileChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = e.target;
    if (!files) return;

    const newPhotos = await Promise.all(
      Array.from(files).map(async (file) => {
        const reader = new FileReader();
        return new Promise<string>((resolve) => {
          reader.onloadend = () => {
            resolve(reader.result as string);
          };
          reader.readAsDataURL(file);
        });
      })
    );

    setPhotos((prevPhotos) => [...prevPhotos, ...newPhotos]);
    const previousPhotos = getValues("imageUrls");
    setValue("imageUrls", [...(previousPhotos || []), ...newPhotos]);
  };

  const removePhoto = (index: number) => {
    setPhotos(photos.filter((_, idx) => idx !== index));
  };

  const onSubmit = async ({ imageUrls, title, text }: FormValuesProps) => {
    try {
      if (type === "image") {
        dispatch(addPost("image", imageUrls || [], title));
      } else {
        dispatch(addPost("text", text || "", title));
      }
      if (closeModal) closeModal();
    } catch (error) {
      console.error("Failed to submit application:", error);
    }
  };

  return (
    <FormProvider methods={methods} onSubmit={handleSubmit(onSubmit)}>
      <section className="flex flex-col gap-2 mt-2">
        <RHFTextArea name="title" placeholder="Title" maxLength={300} />
        {type === "text" && (
          <RHFTextArea name="text" placeholder="Body" className="h-40" />
        )}
        {type === "image" && photos.length <= 0 && (
          <FileUploadArea handleFileChange={handleFileChange} />
        )}
        {type === "image" && photos.length > 0 && (
          <PostPhotoCarousel
            handleFileChange={handleFileChange}
            photos={photos}
          />
        )}

        <div className="flex gap-5 justify-end">
          <Button className="rounded-full bg-blue-800" type="button">
            Save draft
          </Button>
          <Button type="submit" className="rounded-full bg-blue-800">
            Post
          </Button>
        </div>
      </section>
    </FormProvider>
  );
}
