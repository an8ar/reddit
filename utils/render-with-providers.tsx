import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { NextIntlClientProvider, IntlConfig } from 'next-intl';
import { AppStore, RootState, makeStore } from '~/store';

interface IntlProviderProps extends Omit<IntlConfig, 'locale'> {
  locale: string;
}

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
  intlProviderProps?: IntlProviderProps;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = makeStore(),
    intlProviderProps,
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <NextIntlClientProvider {...intlProviderProps}>{children}</NextIntlClientProvider>
    </Provider>
  );

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
