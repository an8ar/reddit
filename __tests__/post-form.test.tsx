import React from 'react';
import { screen, fireEvent, act } from '@testing-library/react';
import { renderWithProviders } from '../utils/render-with-providers';
import { PostForm } from '~/feature/posts/components/post-form';
import enMessages from '../messages/en.json';
import ruMessages from '../messages/ru.json';
import '@testing-library/jest-dom';
import { useRouter } from 'next/navigation';
import { makeStore } from '~/store';

jest.mock('next/navigation', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const mockRouter = {
  push: jest.fn(),
};

describe('PostForm Component', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  test('renders correctly with initial content', () => {
    const { container } = renderWithProviders(<PostForm type="image" />, {
      intlProviderProps: {
        locale: 'en',
        messages: enMessages,
        timeZone: 'America/New_York',
        now: new Date(),
      },
    });
    expect(screen.getByPlaceholderText('Title')).toBeInTheDocument();

    expect(screen.getByRole('button', { name: 'Post' })).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('renders correctly with different locale', () => {
    const { container } = renderWithProviders(<PostForm type="image" />, {
      intlProviderProps: {
        locale: 'ru',
        messages: ruMessages,
        timeZone: 'America/New_York',
        now: new Date(),
      },
    });

    expect(screen.getByPlaceholderText('Заголовок')).toBeInTheDocument();

    expect(screen.getByRole('button', { name: 'Добавить' })).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('handles form submission', async () => {
    const store = makeStore();

    store.dispatch = jest.fn();

    await act(async () => {
      renderWithProviders(<PostForm type="image" />, {
        store,
        intlProviderProps: {
          locale: 'en',
          messages: enMessages,
          timeZone: 'America/New_York',
          now: new Date(),
        },
      });
    });

    const expectedUpdateFormPayload = {
      title: 'Test Title',
    };

    const expectedAddPostPayload = {
      id: expect.any(String),
      createdAt: expect.any(String),
      title: 'Test Title',
      imageUrls: [],
      voteCount: 0,
    };

    await act(async () => {
      fireEvent.change(screen.getByPlaceholderText('Title'), {
        target: { value: 'Test Title' },
      });
    });

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'postSlice/updateForm',
      payload: expectedUpdateFormPayload,
    });

    await act(async () => {
      fireEvent.submit(screen.getByRole('button', { name: 'Post' }));
    });

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'postSlice/addPost',
      payload: expectedAddPostPayload,
    });

    expect(mockRouter.push).toHaveBeenCalledWith('/');
  });
});
