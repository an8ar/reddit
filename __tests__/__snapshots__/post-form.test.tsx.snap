// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post form rendering 1`] = `
<div>
  <form>
    <section
      class="flex flex-col gap-2 mt-2"
    >
      <div
        class="space-y-2 relative"
      >
        <textarea
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 text-base"
          id=":r0:-form-item"
          maxlength="300"
          name="title"
          placeholder="Title"
        />
        <div
          class="text-right text-sm text-gray-500"
        >
          /
          300
        </div>
      </div>
      <div
        class="relative w-full aspect-video rounded-2xl border-2 border-dotted border-slate-600  hover:cursor-pointer h-40 dot"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 absolute inset-0 flex gap-4 items-center justify-center"
          for="inputImage"
        >
          <span />
          <span
            class="text-lg"
          >
            Add Photo
          </span>
        </label>
        <input
          accept="image/*"
          class="h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 file-selector-button-none hidden"
          hidden=""
          id="inputImage"
          multiple=""
          type="file"
        />
      </div>
      <div
        class="flex gap-5 justify-end"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 rounded-full bg-blue-800"
          type="button"
        >
          Save draft
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 rounded-full bg-blue-800"
          type="submit"
        >
          Post
        </button>
      </div>
    </section>
  </form>
</div>
`;
